applications/IgaApplication/custom_utilities/node_curve_geometry_3d.h

```diff
@@ -0,0 +1,206 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Thomas Oberbichler
+//
+
+#if !defined(KRATOS_NODE_CURVE_GEOMETRY_3D_H_INCLUDED)
+#define KRATOS_NODE_CURVE_GEOMETRY_3D_H_INCLUDED
+
+// System includes
+
+// External includes
+#include <ANurbs/Core>
+
+// Project includes
+#include "includes/define.h"
+#include "includes/Node.h"
+#include "includes/variables.h"
+
+namespace Kratos {
+
+/// Spatial NURBS curve geometry with Kratos Nodes as control points.
+/** Unlike the ANurbs::CurveGeometry, this class does not use static Points as
+ *  control points but Finite Element Nodes. That means that the Geometry
+ *  changes whenever the Nodes are moving.
+ */
+class NodeCurveGeometry3D
+    : public ANurbs::CurveGeometryBase<double, Kratos::array_1d<double, 3>>
+{
+protected:
+    using NodePointer = typename Node<3>::Pointer;
+
+public:
+    using CurveGeometryBaseType = ANurbs::CurveGeometryBase<double,
+        Kratos::array_1d<double, 3>>;
+    using typename CurveGeometryBaseType::KnotsType;
+    using typename CurveGeometryBaseType::ScalarType;
+    using typename CurveGeometryBaseType::VectorType;
+ 
+protected:
+    std::vector<NodePointer> mNodes;
+
+public:
+    /** Creates a new NodeCurveGeometry3D.
+     *
+     *  \param Degree Degree of the curve
+     *  \param NumberOfNodes Number of nodes
+     */
+    NodeCurveGeometry3D(
+        const int Degree,
+        const int NumberOfNodes)
+        : CurveGeometryBaseType(Degree, NumberOfNodes)
+        , mNodes(NumberOfNodes)
+    {
+    }
+
+    /** Gets the Kratos node at a given index.
+     * 
+     * \param Index Index of the node
+     * 
+     * @return Kratos node at the given index.
+     */
+    NodePointer Node(
+        const int Index
+    ) const
+    {
+        KRATOS_DEBUG_ERROR_IF(Index < 0 || NbPoles() <= Index) <<
+            "Index out of range" << std::endl;
+
+        return mNodes[Index];
+    }
+
+    /** Sets the Kratos node at a given index.
+     * 
+     * \param Index Index of the node
+     */
+    void SetNode(
+        const int Index,
+        NodePointer Value
+    )
+    {
+        KRATOS_DEBUG_ERROR_IF(Index < 0 || NbPoles() <= Index) <<
+            "Index out of range" << std::endl;
+
+        mNodes[Index] = Value;
+    }
+
+    /** Gets the location of the Kratos node at a given index.
+     * 
+     * \param Index Index of the node
+     * 
+     * @return Location of the Kratos node at the given index.
+     */
+    VectorType Pole(
+        const int Index) const override
+    {
+        const auto& node = *Node(Index);
```

Comment:
uuuuuuf, I don't like very much thsi auto with & and a *. Looks like an atomic boom

Reply:
total agree...